### ISSUE TITLE
Max retries exceeded

### ISSUE THREAD

**Issue by Pascal31500** [theneweinstein/somneo#66]
After a certain time (variable) the Somneo control no longer works, see error message:

Failed to perform the action light/turn_on. HTTPSConnectionPool(host='xxx.xxx.xxx.xxx', port=443): Max retries exceeded with url: /di/v1/products/1/wulgt (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f66e32900>: Failed to establish a new connection: [Errno 111] Connection refused'))

**Comment by Pascal31500** [comment]
r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/requests/adapters.py", line 700, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='xxx.xxx.xxx.xxx', port=443): Max retries exceeded with url: /di/v1/products/1/wulgt (Caused by NewConnectionError(': Failed to establish a new connection: [Errno 111] Connection refused'))

**Comment by Eikei7** [comment]
I get pretty much the same thing.

**Comment by Eikei7** [comment]
Or rather, I get this when I try to set it up:

HTTPSConnectionPool(host='192.168.1.223', port=443): Max retries exceeded with url: /upnp/description.xml (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f1f122517f0>: Failed to establish a new connection: [Errno 111] Connection refused'))

**Comment by drewbern** [comment]
I noticed that unplugging the device for a few seconds usually solves this issue, for a short time at least. I also noticed that when this error pops up, I am unable to establish a connection with the SleepMapper app as well. Not sure if the state is cause by something in this repo which then prevents normal communication via SleepMapper, or if this is an issue with the device software itself.

**Comment by skuijper** [comment]
Yes, this looks very much like the issue I was having. I suspect it has to do with poor wifi reception combined with the inability to handle less than optimal wifi reception. Restarting it always makes it connect, but if after a while the reception has dropped (for a moment or maybe after a certain amount of times) the device becomes inaccessible. Even though it indicates that it is connected on the device display. 

Since it had worked before without changing anything I suspect the firmware is the culprit. Maybe it just stalls in these type of (edge) cases.

My solution was to improve wifi signal by rearranging my wifi mesh devices (I seemed to have some kind of black hole around my nightstand).

Still this is not 100% failsafe (although the device usually keeps on working a.t.m.) because it may still stop responding after any kind of wifi hick-up. That's why I also installed a smart power socket together with an automation which detects unavailability. It just power cycles the device when it becomes unavailable.

**Comment by jonatino** [comment]
I have this as well. Does the somneo do rate limiting on its API's? I wonder if different poll times avoid this issue. Because even when the HA device fails to connect, you can still ping the ip normally and view it in a web browser.

**Comment by MickazNL** [comment]
Just wanted to say I experience the same issue. 

I think an update on the device itself has changed things.

**Comment by matthiasengelhardt77** [comment]
I believe the issue comes from the device firmware. I had problems to connect the official _SleepMapper_ app already before I used this integration. I switch off the WIFI access point in my sleeping room every night. Sadly, it seems that the _Somneo_ device cannot re-connect to the network without a power-cycle. :-(

**Comment by SchwurbeI** [comment]
My Somnbeo only accepts one connection at a time, and if the client goes away without closing it properly (e.g. an HA restart), the device never reaps the half-open socket. Seems like client side sits in `FIN_WAIT_2` and it stays unreachable on both 443 and 80 until the device is power cycled.

As far as I can tell nothing in the integration or pysomneo ever closes the session, so every HA restart strands one. I patched mine to send a final request with `Connection: close` on `EVENT_HOMEASSISTANT_STOP`, and now it reconnects after a restart without a power cycle.

Would a `close()` in pysomneo plus a stop hook in the integration make sense?

**Comment by skuijper** [comment]
Just some extra info: I'm using Homey now, with this integration: https://homey.app/en-nl/app/com.philips.somneo/Philips-Sleepmapper/

Gives the same behaviour. Could very well be that it uses the same pysomneo lib behind the curtains. Can't really see, because the source code is not available.

If not, it's either the same unclosed connection implementation as with pysomneo or something else...

**Comment by theneweinstein** [comment]
Thanks for the suggestions, I'm currently working to improve the integration and underlying libary. This is going slowly, but I hope it solves a few of these issues.
